library/network/src/modules/SuSEFirewallServices.rb

```diff
@@ -745,6 +780,157 @@ def GetPossiblyConflictServices
     publish function: :GetPossiblyConflictServices, type: "list <string> ()"
   end
 
-  SuSEFirewallServices = SuSEFirewallServicesClass.new
-  SuSEFirewallServices.main
+  class SuSEFirewalldServicesClass < SuSEFirewallServicesClass
+    include Yast::Logger
+
+    SERVICES_DIR = ["/etc/firewalld/services", "/usr/lib/firewalld/services"]
+
+    IGNORED_SERVICES = ["..", "."]
+
+    def initialize
+      @services = nil
+
+      @known_services_features = {
+        "TCP"     => "tcp_ports",
+        "UDP"     => "udp_ports",
+        "IP"      => "ip_protocols",
+        "MODULES" => "modules"
+      }
+
+      @known_metadata = { "Name" => "name", "Description" => "description" }
+
+      # firewall needs restarting. Always false for firewalld
+      @sfws_modified = false
+    end
+
+    # Reads services that can be used in FirewallD
+    # @note Contrary to SF2 we do not read the full service details here
+    # @note since that would mean to issue 5-6 API calls for every service
+    # @note file which will take a lot of time for no particular reason.
+    # @note We will read the full service information if needed in the
+    # @note service_details method.
+    # @return [Boolean] if successful
+    def ReadServicesDefinedByRPMPackages
+      log.info "Reading FirewallD services from #{SERVICES_DIR.join(" and ")}"
+
+      @services ||= {}
+
+      SuSEFirewall.api.services.each do |service_name|
+        # Init everything
+        @services[service_name] = {}
+        @known_services_features.merge!(@known_metadata).each do |_, v|
```

Comment:
why `merge!` variant is needed and not using simple `merge`? it will have nasty side-effect of modifying `@known_services_features` which is not nice.
and btw there is `each_value` method. So it can look like
`@known_services_features.merge(@known_metadata).each_value do |v|`
btw v is not nice for multiline block, what value stands for? metadata type?